### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by viktorprogger**
Trying to get this package working

### CODE REVIEW

**Comment by viktorprogger** [comment]
I've finished with preview version. What is still not done:
- Code style
- Drivers (interface is ready, drivers will be implemented in other repos)
- Cli commands
- Tests
- Not all logs are moved from `LogBehavior`

Basic concept is finished as for me. The only things I want to rethink are events.

File: src/Queue.php [hunk 1]
```diff
@@ -103,7 +108,7 @@ public function ttr(int $value): self
      *
      * @return $this
      */
-    public function delay(int $value): self
+    public function withDelay(int $value): self
```
Comment (viktorprogger) [inline src/Queue.php]: WDYT about moving `withDelay`, `withPriority` and `ttr` to different Job interfaces? We will can make `canPush` in the `DriverInterface`, but I'm not sure about use cases for that.

File: composer.json [hunk 2]
```diff
@@ -23,23 +23,28 @@
...
         "yiisoft/serializer": "^3.0@dev",
-        "symfony/process": "^4.0",
-        "php": "^7.4"
+        "yiisoft/strings": "^3.0@dev"
     },
     "require-dev": {
+        "ext-pdo": "*",
+
+        "aws/aws-sdk-php": ">=2.4",
         "enqueue/amqp-lib": "^0.9",
```
Comment (samdark) [inline composer.json]: It is used in `src`. Likely it worth moving it and all related code to separate driver-package later.

File: src/Jobs/AttemptsRestrictedJob.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Queue\Jobs;
+
+use Throwable;
+
+abstract class AttemptsRestrictedJob implements RetryableJobInterface
+{
+    protected int $attemptsMax;
```
Comment (samdark) [inline src/Jobs/AttemptsRestrictedJob.php]: These should be initialized.

File: src/Processors/JobProcessor.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Throwable;
+use Yiisoft\Log\Logger;
+use Yiisoft\Yii\Queue\Events\ExecEvent;
+use Yiisoft\Yii\Queue\LogMessageFormatter;
+use Yiisoft\Yii\Queue\MessageInterface;
+use Yiisoft\Yii\Queue\Queue;
+
+class JobProcessor implements JobProcessorInterface
```
Comment (samdark) [inline src/Processors/JobProcessor.php]: These are usually called "workers".

File: src/Queue.php [hunk 5]
```diff
@@ -131,171 +136,61 @@ public function priority($value): self
...
-     */
-    public function isDone($id): bool
+    public function listen(): void
     {
-        return $this->status($id) === self::STATUS_DONE;
+        $pid = getmypid();
+        $message = "Worker $pid is started.";
+        $this->logger->info($message);
+
+        $handler = static function (MessageInterface $message) {
```
Comment (samdark) [inline src/Queue.php]: Should not be `static` since there's `$this` used in it.

File: tests/App/config/console.php [hunk 6]
```diff
@@ -1,33 +1,23 @@
...
-                'db' => 'pgsql',
-                'migrationPath' => null,
-                'migrationNamespaces' => [
-                    'Yiisoft\Yii\Queue\Drivers\Db\migrations',
-                ],
-            ],
-            'benchmark' => \Yiisoft\Yii\Queue\Tests\App\Benchmark\Controller::class,
-        ],
-    ],
+    Application::class => static function (ContainerInterface $container) {
```
Comment (samdark) [inline tests/App/config/console.php]: That would likely mess with other console commands. Can we do it like here? https://github.com/yiisoft/yii-demo/blob/master/config/params.php#L27

File: src/Queue.php (hunk 1 above)
Comment (samdark) [reply src/Queue.php]: What do you mean by "different" job interfaces?

File: composer.json (hunk 2 above)
Comment (viktorprogger) [reply composer.json]: Yes, all drivers will be implemented separately, as I've already said above.

File: tests/App/config/console.php (hunk 6 above)
Comment (viktorprogger) [reply tests/App/config/console.php]: I didn't touch any test and its environment, except a couple of first commits and automatic refactoring. If you are ok with the basic archetecture I will adjust tests to it and to `yii-console`.

File: src/Queue.php (hunk 1 above)
Comment (viktorprogger) [reply src/Queue.php]: - `DelayableJobInterface`
- `PrioritisedJobInterface`
- `ReservableJobInterface`

Comment (samdark) [reply src/Queue.php]: That makes sense only if these will be used separately. Would they?

Comment (viktorprogger) [reply src/Queue.php]: Not all drivers will support at least delayable and prioritised jobs. Will check the same for job reservation tomorrow or next week.

File: src/Events/JobEvent.php [hunk 7]
```diff
@@ -18,28 +20,29 @@
...
      */
-    public $sender;
+    public Queue $sender;
     /**
      * @var string
      */
-    public $name;
+    public string $name;
     /**
      * @var string|null unique id of a job
```
Comment (xepozz) [inline src/Events/JobEvent.php]:
```suggestion
     * @var string unique id of a job
```

File: src/Events/JobEvent.php [hunk 8]
```diff
@@ -18,28 +20,29 @@
 abstract class JobEvent
 {
     /**
-     * @var \Yiisoft\Yii\Queue\Queue
+     * @var Queue
      */
-    public $sender;
+    public Queue $sender;
```
Comment (xepozz) [inline src/Events/JobEvent.php]: Make getter for it

File: src/Events/JobEvent.php [hunk 9]
```diff
@@ -18,28 +20,29 @@
...
     /**
      * @var string
      */
-    public $name;
+    public string $name;
```
Comment (xepozz) [inline src/Events/JobEvent.php]: Make getter for it

File: src/Events/JobEvent.php [hunk 10]
```diff
@@ -18,28 +20,29 @@
...
     /**
      * @var string|null unique id of a job
      */
-    public $id;
+    public ?string $id;
```
[the remark this answers is not in the document]
Comment (xepozz) [reply src/Events/JobEvent.php]: Make getter for it

File: src/Events/JobEvent.php [hunk 11]
```diff
@@ -18,28 +20,29 @@
...
     /**
-     * @var \Yiisoft\Yii\Queue\JobInterface
+     * @var JobInterface|RetryableJobInterface
      */
-    public $job;
+    public ?JobInterface $job;
```
Comment (xepozz) [inline src/Events/JobEvent.php]: Make getter for it

Comment (xepozz) [reply src/Events/JobEvent.php]: Is it needs to be public?

File: src/Events/JobEvent.php [hunk 12]
```diff
@@ -18,28 +20,29 @@
...
     /**
      * @var int time to reserve in seconds of the job
      */
-    public $ttr;
+    public int $ttr;
```
Comment (xepozz) [inline src/Events/JobEvent.php]: Make getter for it

File: src/Events/JobEvent.php [hunk 13]
```diff
@@ -18,28 +20,29 @@
...
 
-    public function __construct(string $name, $id, JobInterface $job, int $ttr)
+    public function __construct(string $name, string $id, ?JobInterface $job, int $ttr)
```
Comment (xepozz) [inline src/Events/JobEvent.php]: Why the `$job` is nullable?

File: src/Queue.php [hunk 14]
```diff
@@ -131,171 +136,61 @@ public function priority($value): self
...
     {
-        return $this->status($id) === self::STATUS_DONE;
+        $pid = getmypid();
+        $message = "Worker $pid is started.";
+        $this->logger->info($message);
+
+        $handler = function (MessageInterface $message) {
+            try {
+                $this->processor->process($message, $this);
+            } catch (Throwable $exception) {
```
[the remark this answers is not in the document]
Comment (viktorprogger) [reply src/Queue.php]: I don't remember why exactly I made this. I think it was either a mistake or some kind of misunderstanding. Will remove this try/catch in both cases.

File: src/Processors/Worker.php [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+        try {
+            $this->dispatcher->dispatch($event);
+        } catch (Throwable $exception) {
+            $this->handleError($queue, $event, $exception);
+
+            throw $exception;
+        }
+
+        try {
```
[the remark this answers is not in the document]
Comment (viktorprogger) [reply src/Processors/Worker.php]: You are absolutely right. Thanks.

File: src/LogMessageFormatter.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @param ExecEvent $event
+     *
+     * @return string
+     */
+    public function getExecTitle(ExecEvent $event): string
+    {
+        $title = $this->getJobTitle($event);
+        $extra = "attempt: $event->attempt";
```
[the remark this answers is not in the document]
Comment (viktorprogger) [reply src/LogMessageFormatter.php]: It works without these braces too. Or do you mean smthng else?

File: src/LogBehavior.php [hunk 17]
```diff
@@ -148,7 +102,7 @@ protected function getExecTitle(ExecEvent $event): string
     {
         $title = $this->getJobTitle($event);
         $extra = "attempt: $event->attempt";
-        if ($pid = $event->getTarget()->getWorkerPid()) {
+        if ($pid = $event->sender->getWorkerPid()) {
```
[the remark this answers is not in the document]
Comment (viktorprogger) [reply src/LogBehavior.php]: I'm for it. This field must be immutable. But there are some other questions with events I left undone. In fact I have to fully refactor all events and its usages.

**Comment by viktorprogger** [comment]
Made minor fixes today. The issues I'll work on the first of all are:
- See if we need `ReservableJobInterface`
- Redesign events

@samdark, @xepozz, any more high-priority themes to explore?

**Comment by samdark** [comment]
@viktorprogger do you want this PR merged first?

File: src/LogMessageFormatter.php [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public function getJobTitle(JobEvent $event): string
+    {
+        if ($event->job === null) {
+            $name = get_class($event->job);
+        } else {
+            $name = 'unknown job';
+        }
+
+        return "[$event->id] $name";
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/LogMessageFormatter.php]: It is OK as is.

File: src/LogMessageFormatter.php [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    public function getExecTitle(ExecEvent $event): string
```
Comment (samdark) [inline src/LogMessageFormatter.php]: Should be named better. It's not clear what's "exec".

File: src/DriverInterface.php [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Queue;
+
+use Yiisoft\Yii\Queue\Jobs\JobInterface;
+
+interface DriverInterface
+{
+    public function nextMessage(): ?MessageInterface;
```
Comment (samdark) [inline src/DriverInterface.php]: phpdoc needed describing the method.

**Comment by samdark** [comment]
Merged 👍 . Let's continue in separate pull requests.

**Pull request merged**